### DISCUSSION TITLE
Create new virtual ext4 disk and attach to available WSL2

### DISCUSSION THREAD

**Discussion by phongmis** [microsoft/WSL#5896] [General] [answered]
Does WSL2 allow create new ext4 virtual disk and attach to available WSL2 instance ?
If WSL2 allow, how can I do that.

**Comment by guissalustiano** [comment]
just came out of the oven this functionality
https://devblogs.microsoft.com/commandline/access-linux-filesystems-in-windows-and-wsl-2

**Comment by ad-on-is** [comment]
> just came out of the oven this functionality
> https://devblogs.microsoft.com/commandline/access-linux-filesystems-in-windows-and-wsl-2

This blog post explains only how to mount a physical disk, but not how to mount a *.vhdx file.

**Comment by therealkenc** [comment]
> This blog post explains only how to mount a physical disk

Documentation requests and PRs can be sent [here](https://github.com/MicrosoftDocs/WSL).

**Comment by benhillis** [comment, marked as answer]
Yes it does.

**Reply by benapetr** [reply]
This is a terrible answer, he wanted to know HOW and you failed to answer that

**Comment by craigloewen-msft** [comment]
Hey folks, just went through this process and wanted to write it down incase others stumbled upon it.

## How to create a dynamically expanding VHDX with WSL

### Create your base VHD
You can do this multiple ways. I'd recommend going to Hyper-V Manager -> Create New -> Virtual Hard Disk, and then make an expandable one up to whatever size you like. 
![image](https://user-images.githubusercontent.com/42221804/125855589-19ac0425-634d-4f6c-b8fe-45af24aa8d99.png)

### Mount it into WSL using `wsl --mount --bare`
So first mount the VHD into Windows using:
` Write-Output "\\.\PhysicalDrive$((Mount-VHD -Path ./ext4.vhdx -PassThru | Get-Disk).Number)"`
And then mount it into WSL using the output from the above command. For reference mine was: 
`wsl --mount \\.\PhysicalDrive3 --bare` 

### Format the VHD into ext4
You can use the `dmesg` command to see what drive it was mounted as. Mine was mounted under `sdc`. So I use the command: 
` sudo mkfs -t ext4 -i 12000 /dev/sdc`
To format it as ext4 and I've changed the ratio of inodes in it to allow more inodes than normal.

### Mount it in WSL
You can then remount this disk by running
`wsl --unmount; wsl --mount \\.\PhysicalDrive3`
And then you should have access to the disk!

**Reply by edwpang** [reply]
Can I use disk management to create the vhd?

**Reply by ophers** [reply]
yes. see https://winaero.com/create-a-new-vhd-or-vhdx-file-in-windows-10/

**Reply by philipmgrant** [reply]
Unfortunately this doesn't work on stable Windows 10 (because the `--mount` option isn't available): it requires an insider build or Windows 11.

**Reply by ophers** [reply]
You can try installing WSL2 preview from the Store: https://www.microsoft.com/store/productId/9P9TQF7MRM4R
If you hit issues, you can go back simply by uninstalling the store app.
A **major** feature I like is exactly the ability to mount a disk as regular user (does not need elevation), so I can place the command in "startup".

**Reply by josesa-xx** [reply]
WSLg and wsl --mount and other features were backported to Windows 10, you can download it from [KB5020030](https://www.catalog.update.microsoft.com/Search.aspx?q=KB5020030) 

More details on [microsoft devblog](https://devblogs.microsoft.com/commandline/the-windows-subsystem-for-linux-in-the-microsoft-store-is-now-generally-available-on-windows-10-and-11/)

**Comment by anodynos** [comment]
When I use this method to mount an NTFS image onto WSL Linux, I get very low writing speeds of around ~20Mb/Sec for a large file (from Windows to \\wsl.localhost\Ubuntu22.04\mnt\mydisk).

Whereas when I copy the same file inside the main .vhdx  (i.e to my /home/user/) OR when I format the same .vhdx as ext4, I get x10 the speed of around 200mb/sec (on a fast SSD). 

I can't believe just the filesystem (NTFS vs ext4), is making it so slow. NTFS inside Linux works pretty well in general...

**Reply by ophers** [reply]
It's a known issue that accessing from windows into a WSL2 mounted volume is slow.
Assuming you do not need to concurrently access that volume from Windows and WSL2, you could:
1. Shutdown WSL
2. Mount the NTFS vhdx onto your Windows
3. Copy files in/out
4. Dismount the volume from Windows
5. Start the WSL2 instance

**Comment by Thierry-Larmoire** [comment]
on a linux over wsl2 in /mnt/c/xxx, create the image
`qemu-img create -f vhdx storage.vhdx 80G
`
in cmd.exe, expose the image in all wsl2
`wsl --mount storage.vhdx --bare
`
on linux over wsl2, partition it
`sudo fdisk -t gpt /dev/sdx
`
format it
`sudo mkfs.ext4 /dev/sdx1
`
in cmd.exe, unexpose, expose with mounting the partition
`wsl --unmount storage.vhdx
`
`wsl --mount storage.vhdx --name data --partition 1
`
voilà
